keps/sig-scheduling/20190221-even-pods-spreading.md

```diff
@@ -0,0 +1,350 @@
+  is a tool that automatically adjusts the size of the Kubernetes cluster upon
+  specific conditions.
+
+## Summary
+
+`EvenPodsSpreading` feature gives users more fine-grained control on
+distribution of pods scheduling, so as to achieve better high availability and
+resource utilization.
+
+## Motivation
+
+In Kubernetes, "Affinity" related directives are aimed to control how pods are
+scheduled - more packed or more scattering. But right now only limited options
+are offered: for `PodAffinity`, infinite pods can be stacked onto qualifying
+topology domain(s); for `PodAntiAffinity`, pods are scheduled onto each single
+topology domain exclusively.
+
+This is not an ideal situation if users want to put pods evenly across different
+topology domains - for the sake of high availability or saving cost. And regular
+rolling upgrade or scaling out replicas can also be problematic. See more
+details in [user stories](#user-stories).
+
+> Even pods spreading is a long-discussed topic, but it's only feasible to
+> hammer out the design/implementation details after recent Affinity performance
+> improvements.
+
+### Goals
+
+- Even spreading is calculated among pods instead of apps API (such as
+  Deployment, ReplicaSet).
+- Even spreading can be either a predicate (hard requirement) or a priority
+  (soft requirement).
+- Even spreading _might_ be implemented on limited topologies in initial
+  version.
+
+### Non-Goals
+
+- Even spreading is NOT calculated on an application basis. In other words, it's
+  not _only_ applied within replicas of an application.
+- "Max number of pods per topology" is NOT a goal.
+- Scale-down on an application is not guaranteed to achieve even pods spreading
+  in initial implementation.
+
+## Proposal
+
+### User Stories
+
+#### Story 1 - NodeAffinity
+
+As an application developer, I want my application pods to be scheduled onto
+specific topology domains (via NodeAffinity with spec "app in [zone1, zone2,
+zone3]"), but I don't want them to be stacked too much on one topology. (see
+[#68981](https://github.com/kubernetes/kubernetes/issues/68981))
+
+#### Story 2 - PodAffinity
+
+As an application developer, I want my application pods to co-exist with
+particular pods in the same topology domain (via PodAffnity), and I want them to
+be deployed onto separate nodes (or sub domains) as even as possible.
+
+#### Story 3 - PodAntiAffinity
+
+As an application developer, I want my application pods not to co-exist with
+specific pods (via PodAntiAffinity). But in some cases it'd be favorable to
+tolerate "violating" pods in a manageable way. For example, suppose an app
+(replicas=2) is using PodAntiAffinity and deployed onto a 2-nodes cluster, and
+next the app needs to perform a rolling upgrade, then a third replacement pod is
+created, but it failed to be placed due to lack of resource. In this case,
+
+- if CA is enabled, a new machine will be provisioned to hold the new pod
```

Comment:
what is CA ?

Reply:
https://github.com/kubernetes/enhancements/pull/851/files#diff-31efc10ab6301e5b166ef21358b07644R53